p2p/simulations/network_test.go

```diff
@@ -18,14 +18,248 @@ package simulations
+		t.Fatal(err)
+	}
+	log.Debug("snapshot taken", "nodes", len(snap.Nodes), "conns", len(snap.Conns), "json", string(j))
+
+	// verify that the snap element numbers check out
+	if len(checkIds) != len(snap.Conns) || len(checkIds) != len(snap.Nodes) {
+		t.Fatalf("snapshot wrong node,conn counts %d,%d != %d", len(snap.Nodes), len(snap.Conns), len(checkIds))
+	}
+
+	// shut down sim network
+	runningOne = false
+	sub.Unsubscribe()
+	network.Shutdown()
+
+	// check that we have all the expected connections in the snapshot
+	for nodid, nodConns := range checkIds {
+		for _, nodConn := range nodConns {
+			var match bool
+			for _, snapConn := range snap.Conns {
+				if snapConn.One == nodid && snapConn.Other == nodConn {
+					match = true
+					break
+				} else if snapConn.Other == nodid && snapConn.One == nodConn {
+					match = true
+					break
+				}
+			}
+			if !match {
+				t.Fatalf("snapshot missing conn %v -> %v", nodid, nodConn)
+			}
+		}
+	}
+	log.Info("snapshot checked")
+
+	// PART II
+	// load snapshot and verify that exactly same connections are formed
+
+	adapter = adapters.NewSimAdapter(adapters.Services{
+		"noopwoop": func(ctx *adapters.ServiceContext) (node.Service, error) {
+			return NewNoopService(nil), nil
+		},
+	})
+	network = NewNetwork(adapter, &NetworkConfig{
+		DefaultService: "noopwoop",
+	})
+	defer func() {
+		network.Shutdown()
+	}()
+
+	// subscribe to peer events
+	// every node up and conn up event will generate one additional control event
+	// therefore multiply the count by two
+	evC = make(chan *Event)
+	sub = network.Events().Subscribe(evC)
+	defer sub.Unsubscribe()
+
+	// Channel that is signalling when the Load function returns
+	// to fail the test in the event for loop if that happens
+	// before all connection events are counted.
+	loadDoneC := make(chan struct{})
+
+	go func() {
+		// collect connection events up to expected number
+		ctx, cancel = context.WithTimeout(context.TODO(), time.Second*3)
+		defer cancel()
+
+		connEventCount = nodeCount
+
+	OUTER_TWO:
+		for {
+			select {
+			case <-ctx.Done():
+				t.Fatal(ctx.Err())
+			case ev := <-evC:
+				if ev.Type == EventTypeConn && !ev.Control {
+
+					// fail on any disconnect
+					if !ev.Conn.Up {
+						t.Fatalf("unexpected disconnect: %v -> %v", ev.Conn.One, ev.Conn.Other)
+					}
+					log.Debug("conn", "on", ev.Conn.One, "other", ev.Conn.Other)
+					checkIds[ev.Conn.One] = append(checkIds[ev.Conn.One], ev.Conn.Other)
+					checkIds[ev.Conn.Other] = append(checkIds[ev.Conn.Other], ev.Conn.One)
+					connEventCount--
+					log.Debug("ev", "count", connEventCount)
+					if connEventCount == 0 {
+						break OUTER_TWO
+					}
+				}
+			case <-loadDoneC:
+				// if load function returns before all expected events are caught.
+				// fail the test
+				t.Fatal("load function returned before all connections are established")
+			}
+		}
+
+		// check that we have all expected connections in the network
+		for _, snapConn := range snap.Conns {
```

Comment:
@janos since this is now a separate goroutine, can we be sure all of this will be executed before the test ends? The main routine may finish first...